fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/api/lookup/v1/item/ItemKey.java

```diff
@@ -31,7 +32,10 @@
 
 /**
  * The immutable combination of an item and additional NBT data. Compare using {@link ItemKey#equals}.
+ *
+ * <p>Do not implement.
  */
+@ApiStatus.NonExtendable
 public interface ItemKey {
 	ItemKey EMPTY = ItemKeyImpl.of(Items.AIR, null);
```

Comment:
Any reason of keeping a separate ItemKeyImpl? Imo you can just have this as a pojo than writing a separate impl

Reply:
I'd agree here with a final class pojo for this.

Reply:
I think it should be an interface, to allow people to make copy-on-write ItemKeys and whatnot

Reply:
Item keys are supposed to be immutable and thus thread safe, what the heck are you suggesting @devan-kerman
also if this is an interface, it is hard to implement equals across interfaces

Reply:
What? the ItemKey itself can be immutable, the underlying nbt tag is copied when the ItemKey is initialized, which isn't always necessary, people should be allowed to make their own ItemKey implementations, because they can optimize, like that case. Another one would be if some tool's ItemKey only cares about the durability, instead of storing a CompoundTag, it can store an int (if it doesn't have additional data of course)

Reply:
Then how does the equals logic between different item keys work? It doesn't, and the item key interface has no point of existence. just make it a concrete class

Reply:
??? what, just compare the compound tag and item in the equals method. Our special ItemKey would just test if the other ItemKey had some specific data in it by grabbing its tag.

Reply:
I don't think this should be a pojo. The `ItemKey`s will be cached to improve performance as much as possible, and the hash code is cached too. All those impl details should not be exposed to the API user. We can probably move all the default implementations to the impl though, as they clutter the API class a lot for no reason.

As for custom implementations of an item key... Item keys can be passed to other threads or stored directly in any container without further notice, and that introducing any way to mutate them is the worst possible thing we could do besides using `ItemStack` directly...

Reply:
> just compare the compound tag and item in the equals method
@devan-kerman you realize this just defeats your whole point of tool not having to keep a full stack for performance purposes?
In the end the nbt isn't too big a hurdle. Your "save a tag" is pretty worthless compared to the problems and risks it introduces. if tags are that expensive why does mojang add a nbt to every tool item?

Reply:
> I don't think this should be a pojo

Being a pojo doesn't prevent it from being cached. You will just hide the constructor to be package private and users can only call a static method to obtain instances.

Also since there is mutability concern, I suggest making ItemKey a public final class to ensure safety.

Reply:
I am not sure I understand the value of making this a pojo? I fail to see a direct benefit, and I would very much prefer the ugly implementation details to be kept outside of public API classes.

Reply:
> you realize this just defeats your whole point of tool not having to keep a full stack for performance purposes?

no? copying 2 tags != copying 1 tag, and u could also short cut in some cases too.

> if tags are that expensive why does mojang add a nbt to every tool item?

they also made TACS, item performance can actually be a pretty big problem

Reply:
I don't think Mojang's decisions are made with modded item transfer in mind...

Anyway, I don't think there is a viable alternative to the `ItemKey` concept as it currently stands. You can't reasonably have immutable item + NBT, and allow mutating the NBT at the same time. Do however note that a lot of modded containers or pipes never touch the NBT directly. In some cases this may cause a tiny loss of performance, but I expect chest performance to be much better with `ItemKey`s because the stacks don't have to be copied... Splitting an expensive NBT stack of 2 items into 2 chests does not involve any allocation with item keys.